lagoueduCol/iOS-linyongjian
# Add SwiftUI Support

## Summary

- Github issue/doc: https://github.com/lagoueduCol/iOS-linyongjian/issues/12
- Card: N/A

Use SwiftUI to replace the existing UIKit implementation for the veiw layer.

## Details
### Description
1. Add a feature toggle to turn SwiftUI on or off: `InternalToggle.isSwiftUIEnabled` and let the internal users configure it on the Internal app.
2. Add `Kingfisher/SwiftUI` to support async image loading.
3. Start the SwiftUI implementation in `SceneDelegate` when the Toggle  `InternalToggle.isSwiftUIEnabled`  is on.
4. Implement `SwiftUIMomentsListView` for the Moments List Screen including the user profile and moment.
5. Use `SwiftUIMomentsListItemView` to conditionally return either `SwiftUIUserProfileListItemView` or `SwiftUIMomentListItemView`
6. Implement `SwiftUIUserProfileListItemView` for showing the user profile
7. Implement `SwiftUIMomentListItemView` for showing the moment
8. Implement `LikeToggleStyle` for a neumorphic design toggle for Like.
9. Implement `BlurView` using UIKit's `UIVisualEffectView`
10. Use `MomentsListObservableObject` to bridge RxSwift and Combine


### Screengrabs (if applicable)
![Moments](https://user-images.githubusercontent.com/573856/99899633-f0e84000-2cfe-11eb-9e7e-849a902f10de.gif)

## Quality Analysis

- [ ] Unit tests that cover all added and changed code
- [ ] Tested on Simulator
- [ ] Tested on iPhone
- [ ] Tested on iPad (if applicable) - N/A

**Testing steps:**

0. Install the Internal app from Firebase Distribution app and open it.
0. Shake the phone and activate the Internal menu.
0. Turn on the Future Toggle `SwiftUI Enable`
0. Kill the app (you can tap on `Crash App` to kill it 💥)
0. Re-open the app, should see the SwiftUI
0. You should be able to scroll up and down in the moments list screen
0. You should be able to drag the user profile(Jake Lin with background image and avatar) 
0. You should be able to tap on and off the like button, the liked user list should change according to your operations.

## Checklist

* [x] Has feature toggling been considered?
* [x] Has tested both dark mode and light mode if there is any UI change?
* [x] Has tested Dynamic Type if there is any UI change?
* [x] Has tested language support for multiple locales if there is any UI change?
* [ ] Have new test cases been unit tested? 
* [x] Have run `bundle exec fastlane prepare_pr`?
* [x] Need to labelled the PR? (If applicable: e.g. added new dependencies etc.)
